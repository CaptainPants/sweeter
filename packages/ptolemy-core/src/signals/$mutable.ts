import { MutableValueSignal } from './internal/Signal-implementations/MutableValueSignal.js';
import { type ReadWriteSignal } from './types.js';

// The normal version must be before the optional version for priority reasons.

/**
 * Create a new ReadWriteSignal signal with an initial value.
 * @param initialValue
 * @returns
 */
export function $mutable<T>(initialValue: T): ReadWriteSignal<T>;
/**
 * Create a new ReadWriteSignal signal with a potentially undefined initial value.
 * @param initialValue
 * @returns
 */
export function $mutable<T>(initialValue?: T): ReadWriteSignal<T | undefined>;
/**
 * Create a new ReadWriteSignal signal with an initial value.
 * @param initialValue
 * @returns
 */
export function $mutable<T>(initialValue?: T): ReadWriteSignal<T> {
    if (arguments.length === 0) {
        // arguments.length == 0 means the optional parameter was not passed
        return new MutableValueSignal<T>();
    }
    return new MutableValueSignal<T>(initialValue);
}
