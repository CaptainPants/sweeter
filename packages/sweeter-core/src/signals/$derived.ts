import { MutableCalculatedSignal } from './internal/MutableCalculatedSignal.js';
import { type ReadWriteSignal } from './types.js';

export function $derived<TSource, TKey extends keyof TSource>(
    source: ReadWriteSignal<TSource>,
    key: TKey,
): ReadWriteSignal<TSource[TKey]> & { value: TSource[TKey] } {
    return new MutableCalculatedSignal<TSource[TKey]>(
        () => source.value[key],
        (value) => {
            const sourceObjectOriginal = source.value;

            if (Array.isArray(sourceObjectOriginal)) {
                if (typeof key !== 'number')
                    throw new TypeError(
                        'If source is an array, key must be a number.',
                    );

                const copy = [...sourceObjectOriginal] as TSource;
                copy[key as TKey] = value;
                source.update(copy);
            } else if (
                typeof sourceObjectOriginal === 'object' &&
                sourceObjectOriginal !== null
            ) {
                const copy = { ...sourceObjectOriginal } as TSource;
                copy[key] = value;
                source.update(copy);
            } else {
                throw new TypeError(
                    `Unexpected type ${typeof sourceObjectOriginal}`,
                );
            }
        },
    );
}
