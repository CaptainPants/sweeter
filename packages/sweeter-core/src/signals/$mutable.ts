import { MutableValueSignal } from './internal/MutableValueSignal.js';
import { type ReadWriteSignal } from './types.js';

/**
 * Create a new ReadWriteSignal signal with an initial value.
 * @param initialValue
 * @returns
 */
export function $mutable<T>(initialValue: T): ReadWriteSignal<T> & { value: T };
/**
 * Create a new ReadWriteSignal signal with an undefined initial value.
 * @param initialValue
 * @returns
 */
export function $mutable<TOrUndefined>(): ReadWriteSignal<
    TOrUndefined | undefined
> & { value: TOrUndefined extends undefined ? never : TOrUndefined };
export function $mutable<T>(
    initialValue?: T,
): ReadWriteSignal<T> & { value: T } {
    return new MutableValueSignal<T>(initialValue as T);
}
