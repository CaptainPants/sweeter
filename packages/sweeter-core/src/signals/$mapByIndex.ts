import { type Signal } from '../signals/types.js';
import { type MightBeSignal } from '../types.js';
import { isSignal } from './isSignal.js';
import { $derived } from './$derived.js';
import { $peek, $val, $wrap } from './$val.js';
import {
    addExplicitStrongReference,
    assertNotNullOrUndefined,
} from '@captainpants/sweeter-utilities';
import { $controller } from './$controller.js';
import { SignalState } from './SignalState.js';

export type IndexCacheItem<TMapped> = {
    mappedElement: TMapped;
    detacher: AbortController;
};

export function $mapByIndex<TInput, TMapped>(
    items: MightBeSignal<readonly TInput[]>,
    mappingFun: MightBeSignal<(item: Signal<TInput>, index: number) => TMapped>,
): Signal<readonly TMapped[]> {
    if (!isSignal(items)) {
        // constant array, we can skip a lot of voodoo - the $derived is just because renderItem could be a signal
        return $derived(() =>
            items.map((item, i) => $val(mappingFun)($wrap(item), i)),
        );
    }

    // items is a signal, we need to keep track of a signal for every item
    // including if it changes lengths to dispose/orphan signals that no longer
    // point to a valid index, and add new signals when necessary.

    const cache = $controller(
        SignalState.success<IndexCacheItem<TMapped>[]>([]),
    );

    // store these on an object that is then referenced from the result signal
    const callbacks = {
        reset: () => {
            callbacks.update(true);
        },
        update: (clear = false) => {
            const mappingFunResolved = $peek(mappingFun);
            const updatedInputs = items.peek(); // cloned

            const cacheResolved = cache.signal.peek();
            let updatedCacheResolved: IndexCacheItem<TMapped>[];

            let meaningfullyChanged = false;

            if (clear) { // Act as if the previous cache was empty, and detach every item in the cache
                updatedCacheResolved = [];

                for (const item of cacheResolved) {
                    item.detacher.abort();
                }
            }
            else {
                updatedCacheResolved = [...cacheResolved];

                if (updatedInputs.length < updatedCacheResolved.length) {
                    for (
                        let index = updatedInputs.length;
                        index < updatedCacheResolved.length;
                        ++index
                    ) {
                        const item = updatedCacheResolved[index];
                        assertNotNullOrUndefined(item);
    
                        // Releases the calculated signal (preserving its last value so that
                        // dependencies don't fail in unexpected ways before updating)
                        item.detacher.abort();
                    }
    
                    updatedCacheResolved.length = updatedInputs.length;
                    meaningfullyChanged = true;
                }
            }

            for (let index = 0; index < updatedInputs.length; ++index) {
                let entry = updatedCacheResolved[index];
                if (!entry) {
                    const detach = new AbortController();

                    const input = $derived(() => $val(items)[index] as TInput, {
                        release: detach.signal,
                    });
                    const mapped = mappingFunResolved(input, index);

                    entry = {
                        mappedElement: mapped,
                        detacher: detach,
                    };

                    updatedCacheResolved[index] = entry;
                    meaningfullyChanged = true;
                }
            }

            if (meaningfullyChanged) {
                cache.updateState(SignalState.success(updatedCacheResolved));
            }
        },
    };

    // Initially fill the cache
    callbacks.update();

    if (isSignal(mappingFun)) {
        mappingFun.listenWeak(callbacks.reset);
    }

    items.listenWeak(callbacks.update);

    const resultSignal = $derived(() => {
        return cache.signal.value.map((x) => x.mappedElement);
    });

    addExplicitStrongReference(resultSignal, callbacks);

    return resultSignal;
}
