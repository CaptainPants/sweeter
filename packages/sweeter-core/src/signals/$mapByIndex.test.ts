import { $derived } from './$derived';
import { $mapByIndex } from './$mapByIndex';
import { $mutable } from './$mutable';

it('Does stuff', () => {
    const input = $mutable<string[]>(['alpha', 'beta']);

    const mapped = $mapByIndex(input, (input) => {
        return $derived(() => input.value + '_TEST');
    });

    expect(mapped.peek().map((x) => x.peek())).toMatchSnapshot();

    input.value = ['alpha', 'beta', 'gamma'];

    expect(mapped.peek().map((x) => x.peek())).toMatchSnapshot();
});
