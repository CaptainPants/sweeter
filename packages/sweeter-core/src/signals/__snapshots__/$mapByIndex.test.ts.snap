// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Does stuff 1`] = `
[
  "alpha_TEST",
  "beta_TEST",
]
`;

exports[`Does stuff 2`] = `
[
  "alpha_TEST",
  "beta_TEST",
  "gamma_TEST",
]
`;
